---
title: "HW1"
author: "Ali Chamkalani"
date: "11/09/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## HW 1

VCF (Variant Call Format) is a standardized text file format that is used to store genetic variation information such as SNPs or insertions/deletions (InDel). 
In the following first part of the exercise, we will explore how the information in a VCF is stored. There are some fixed fields (or columns) per record which are explained briefly:

**1. CHROM** - chromosome: The chromosome on which the variant is located, 

**2.POS** - position: The variant (start) position on CHROM

**3. REF** - The reference allele. You do not need to understand what they are for now. Just treat them as some characters.

**4. ALT** - The alternate allele. Same as above.

**5. QUAL** - Quality: Phred-scaled quality score (some notion of the quality for the variant). Higher values represent higher qualities.

**6. INDEL** -insertion/deletion: logical indicating whether variant is an InDel or not.


### !!! The answer to each question is one or couple of lines

# upload data


### Q1: How many InDel variants do you see in this VCF file? Hint: look for INDEL.

numIndel <- which(VCFdata[,9] == TRUE)
> numIndel
  [1]   4   5   9  10  20  24  50  51  55  56  61  64  65  73  79  81  98 102 117 120 131 161 173 177 181 190 195 197 198 203 215 233 238 248 271 274 275 276 278 300 301 308 316 318 324 332 338 343 346 348 353 361 368 370 371
 [56] 375 383 391 403 406 412 415 420 426 432 434 443 453 477 478 484 494 517 524 529 540 549 560 565 588 590 592 616 619 624 627 679 680 694 710 712 720 723 724 754 762 767 795 798 799 801
> sum(numIndel)
[1] 37139
> length(numIndel)
[1] 101

### Q2: Find the position of InDels with read depth (DP filed in the VCF) greater than 15.

> which(VCFdata[,12] > 15)
 [1]  12  14  22  36  37  38  79  93 107 112 120 125 179 187 204 227 240 246 263 273 281 282 283 284 285 286 287 288 294 295 296 299 311 318 330 345 361 367 370 373 390 392 393 427 440 456 458 462 463 468 474 483 497 505 508
[56] 519 541 544 568 587 589 597 598 609 623 634 646 649 650 662 663 664 666 667 668 669 670 671 672 673 675 681 702 724 735 738 757 762 780 794

### Q3: “QUAL” is the phred-scaled quality score, a measure for how likely it is that this site is indeed a variant. Lower QUAL scores indicate low probability of a genetic variation existing at that site. We can convert the confidence value QUAL to a probability value according to the following formula:
### Probability = 1- 10 ^ -(QUAL/10)

### Write a function that takes a VCF file as input and appends the calculated probability value as an additional column to the original VCF file. 
Hint: The input should be the **file name** and output is printing the first 5 lines of the new VCF file

ProbabilityCalculator <- function(filename) {

  dataBase = read.csv(filename, header=TRUE, sep=",")
  
  dataBase$PROB <- 1-10^-(QUAL/10)
  head(dataBase$PROB, 5)
}

###  Q4: Write a function that identifies the frequency of alternate alleles (ALT) for **non indel** variants. In essence your function should take a VCF file as input, and print the frequency (incidence or the number) of ALT alleles. For example, your function will have an output like this (A: 23, C: 5, T: 9, G: 18). In other words, 23 of the records show ALT of A, 5 records show ALT of C, etc

ALT_frequency_function <- function(filename) {

dataBase = read.csv(filename, header=TRUE, sep=",")

non_indel_data <- which(dataBase[,9] == FALSE)
data_of_interest <- c(length(non_indel_data))

for(i in 1:length(non_indel_data)){
counter <-  non_indel_data[i]
data_of_interest[i] <- dataBase[counter,6]
}

ALT_frequency <- table(data_of_interest)
print(ALT_frequency)
}

### Q5: How many variants have more than two deletions? In essence, we need to look at the ALT field and see if the length of the characters in the ALT field is larger than the characters in the REF field (ie ALT > REF +2).  
Print the records that satisfy the above criteria. 
Hint: You would need to use the INDEL, REF, and ALT fields.

ALT_length <- nchar(VCFdata$ALT)
REF_length <- nchar(CDFdata$REF)

print(ALT[which()ALT_length > REF_length + 2])

$
