---
title: "Differential Equations"
author: "Anjali Menon"
---
# Differential Expression Analysis 
```{r setup, include=FALSE}
### Loading Libraries 
knitr::opts_chunk$set(echo = TRUE)

library(reshape)
library(plyr)

library(DESeq2)
library(ggplot2)
library(pheatmap)

library("TCGAbiolinks")
library("AnnotationDbi")

#Following method method for gene annotation was inspired by this website: <https://www.bioconductor.org/help/course-materials/2014/useR2014/Integration.html>
library("org.Hs.eg.db")
library("DESeq2")

#Setting seed for a random value before beginning 
set.seed(800)

### Loading data into R and accessing the Bladder uruthelial carcinoma 

TCGAbiolinks:::getProjectSummary("TCGA-BLCA")
query_TCGA = GDCquery(
  project = "TCGA-BLCA",
  data.category = "Transcriptome Profiling", 
  experimental.strategy = "RNA-Seq",
  workflow.type = "HTSeq - Counts")
result = getResults(query_TCGA)
GDCdownload(query = query_TCGA)
tcga_data = GDCprepare(query_TCGA)

#Finding Relavent genes 
#Accessing the data
mutations <- read.delim("C:/Users/Owner/Desktop/data_mutations.txt", sep = "\t")

top_Genes <- table(mutations$Gene) #Creating a table including all gene information
top_Genes <- sort(top_Genes, decreasing = TRUE) #Ordering genes in descending order

#Plotting genes for visualization
plot(top_Genes, col= "blue")
abline(v=300, col='red')
text(x=3000,y = 200,labels="y=300",col='orange')

#Since the top genes appear to have a frequency of approximately 300 that correlate to that of the mutation, they will be excluded. This is illustrated on the graph using the following line. 

top_Genes <- top_Genes[1:300]
plot(top_Genes, col="blue")

top_Genes <- as.matrix(top_Genes)
top_Genes <- data.matrix(top_Genes)
new_genes <- rownames(top_Genes)
#Creating new table including only the top genes and labeling the columns and rows

top_Genes <- t(rbind(new_genes, top_Genes[, 1]))
colnames(top_Genes) <- c("gene.ID", "mutations")
rownames(top_Genes) <- c(1:length(new_genes))

write.table(new_genes, "new_genes.csv", row.names=FALSE, col.names=FALSE)
write.csv(top_Genes, "top_Genes.csv", row.names=FALSE, col.names=TRUE)

#Organizing data and retrieving gene information
clinical = tcga_data@colData
barcode <- clinical$barcode
patient_age <- clinical$age_at_index
#Ensuring that any value which is not existent or NA is replaced by a zero numerical value
nullval <- which(is.na(patient_age))
patient_age[nullval] = 0
barcode <- t(rbind(barcode, patient_age))
colnames(barcode) <- c("ID", "patient_age")

#Committing the new genes into a new file which is used for further data visualization
new_genes <- read.csv("new_genes.csv", header=FALSE)
new_genes <- new_genes$V1 #Read new genes and assign

#This method was derived from the following website: <https://purrr.tidyverse.org/reference/reduce.html> and it is used to iteratively apply a binary function a and b
find_new <- function(a,b,...){Reduce(intersect,list(a,b,...))}

#Finding new genes that correspond to those in the RNASeq Matrix
g <- find_new(new_genes, rownames(RNAseq_matrix))
for(i in 1:length(g)){
  
  g[i] <- which(rownames(RNAseq_matrix) == g[i])
}
RNAseq_matrix <- RNAseq_matrix[as.numeric(g), ]

#Creating a new data frame including the column names from RNAseq matric and Patient ID's
ID_patient <- colnames(RNAseq_matrix)
ID_patient <- unique(ID_patient)
ID_patient <- find_new(ID_patient, rownames(barcode))

new.data <- as.data.frame(barcode) #Creation of a new data frame containing specific values of interest
column <- colnames(RNAseq_matrix)

samples <- rep(NA, length(new.data$ID)) #Classifying the samples with ID that do not exist
for(i in 1:length(samples)){
  
  samples[i] <- which(column == new.data$ID[i])
}

RNAseq_matrix <- RNAseq_matrix[,samples] #Inserting samples with corresponding length back into original data frame consisting of RNA sequenced data

#STARTING DE ANALYSIS:
new.data$patient_age <- ifelse(new.data$patient_age <= 65, "under 65", "over 65") #These ages were chosen after data visualization as it appeared that there were several patients diagnosed below and above age 65 and we would like to analyse the trends between these 2 age groups by using Differential analysis 

#The following method is referenced from the following website: <https://lashlock.github.io/compbio/R_presentation.html>
#This method was used as DESeq2 can easily use count data to detect differential expressed genes 
differential_Genes = DESeqDataSetFromMatrix(countData=RNAseq_matrix, colData=new.data, design=~patient_age)
differential_Genes = DESeq(differential_Genes)

results.differentialgene <- results(differential_Genes)
res.plot <- as.data.frame(results.differentialgene) #Creating a box plot

ggplot(res.plot, aes(x=log2FoldChange, y=-log(pvalue)))+ geom_point(color = "blue")

#p-value of alpha = 0.05 which means that we are considering values which are statistically significant
alpha.significant <- results(differential_Genes, alpha=0.05)

#Functionality to map feature names of an expression set between common gene ID types such as ENSEMBL and ENTREZ
#<https://www.rdocumentation.org/packages/EnrichmentBrowser/versions/2.2.2/topics/map.ids>
alpha.significant$symbol = mapIds(org.Hs.eg.db, keys=row.names(results.differentialgene), column="SYMBOL", keytype="ENSEMBL", multiVals="first")
alpha.significant$entrez = mapIds(org.Hs.eg.db, keys=row.names(results.differentialgene), column="ENTREZID", keytype="ENSEMBL",
multiVals="first")
alpha.significant$name = mapIds(org.Hs.eg.db, keys=row.names(results.differentialgene),  column="GENENAME", keytype="ENSEMBL",
multiVals="first")

i_val <- (alpha.significant$padj < 0.05)
alpha.significance2 <- subset(alpha.significant, i_val)

#Finding top 5 upregulated and downregulated genes. This procedure was referenced from Assignment 4 to find a heatmap of important genes
upreg <- alpha.significance2[ order( alpha.significance2$log2FoldChange, decreasing=TRUE ), ]
upreg <- upreg[1:5, ]
downreg <- alpha.significance2[ order( alpha.significance2$log2FoldChange ), ]
downreg <- downreg[1:5, ]

#Finding the normalized logarithmic values for creating a heatmap
differential_Genes.new <- estimateSizeFactors(differential_Genes)
differential_Genes.normalized <- counts(differential_Genes.new, normalized=TRUE)
differential_Genes.logFold <- log2(1+differential_Genes.normalized)
numval <- subset(differential_Genes.logFold, i_val)

#Creating a new data frame consisting of both upregulate and downregulated genes 
up <- as.data.frame(upreg$symbol)
colnames(up) <- "Gene.Name"
up_val <- numval[order( alpha.significance2$log2FoldChange, decreasing=TRUE)[1:5], ]
down <- as.data.frame(downreg$symbol)
colnames(down) <- "Gene.Name"
down_val <- numval[order( alpha.significance2$log2FoldChange )[1:5], ]
column.total <- rbind(up, down)
data_genematrix <- as.matrix(rbind(up_val, down_val))
rownames(data_genematrix) <- column.total$Gene.Name

#Creating a heatmap using the clustered values based on the normalized logarithmic values 
column_values = data.frame(new.data$patient_age)
row.names(column_values) <- new.data$ID
pheatmap(data_genematrix, cluster_rows=TRUE, 
         show_rownames=TRUE, 
         show_colnames=FALSE, 
         cluster_cols=FALSE, 
         annotation_col= column_values)
```
